promql/functions.go

```diff
@@ -149,37 +149,108 @@ func extrapolatedRate(vals []parser.Value, args parser.Expressions, enh *EvalNod
 		// of the series, thereby avoiding extrapolation to negative
 		// counter values.
-		// TODO(beorn7): Do this for histograms, too.
-		durationToZero := sampledInterval * (samples.Floats[0].F / resultFloat)
-		if durationToZero < durationToStart {
-			durationToStart = durationToZero
-		}
+		durationToZero = sampledInterval * (samples.Floats[0].F / resultFloat)
+	}
+	if isCounter && resultHistogram != nil && resultHistogram.Count > 0 && len(samples.Histograms) > 0 && samples.Histograms[0].H.Count >= 0 {
+		// Counters cannot be negative. If we have any slope at all
+		// (i.e. resultHistogram.Count went up), we can extrapolate the zero
+		// point of the counter. If the duration to the zero point is shorter
+		// than the durationToStart, we take the zero point as the start
+		// of the series, thereby avoiding extrapolation to negative
+		// counter values.
+		durationToZero = sampledInterval * (samples.Histograms[0].H.Count / resultHistogram.Count)
+	}
+	if durationToZero < durationToStart {
+		durationToStart = durationToZero
 	}
-	extrapolateToInterval += durationToStart
-
 	if durationToEnd >= extrapolationThreshold {
 		durationToEnd = averageDurationBetweenSamples / 2
 	}
-	extrapolateToInterval += durationToEnd
 
-	factor := extrapolateToInterval / sampledInterval
+	if len(samples.Floats) > 0 {
+		extrapolateToInterval += durationToStart + durationToEnd
+
+		factor := extrapolateToInterval / sampledInterval
+		if isRate {
+			factor /= ms.Range.Seconds()
+		}
+
+		resultFloat *= factor
+
+		return append(enh.Out, Sample{F: resultFloat}), annos
+	}
+
+	extrapolateToIntervalRight := extrapolateToInterval + durationToEnd
+	factorRight := extrapolateToIntervalRight / sampledInterval
+
+	factor := (extrapolateToInterval + durationToStart + durationToEnd) / sampledInterval
+
 	if isRate {
 		factor /= ms.Range.Seconds()
+		factorRight /= ms.Range.Seconds()
 	}
-	if resultHistogram == nil {
-		resultFloat *= factor
-	} else {
-		resultHistogram.Mul(factor)
+
+	extrapolateBucket := func(firstValue, resultValue float64) float64 {
+		if firstValue < 0 || resultValue <= 0 {
+			return resultValue * factor
+		}
+		bucketDurationToStart := sampledInterval * (firstValue / resultValue)
+		switch {
+		case bucketDurationToStart < durationToStart:
+			// This bucket extrapolates to zero later (more to the right on the
+			// timeline) than the predicted start, avoid interpolating below
+			// zero.
+			factorLeft := bucketDurationToStart / sampledInterval
+			if isRate {
+				factorRight /= ms.Range.Seconds()
+			}
+			return resultValue*factorLeft + resultValue*factorRight
+		case durationToZero < bucketDurationToStart:
+			// This bucket extrapolates to zero earlier (more to the left on the
+			// timeline) than the overall count, avoid underestimating.
+			// Note if we don't know the duration to Zero, than that's NaN and
+			// this is skipped.
+			compensateLeft := durationToStart / durationToZero
```

Comment:
This strikes me as weird. This is 1 whenever the count is corrected to avoid extrapolation below zero, and it is less than 1 by a certain amount if the count is not corrected at all…